pkg/kubelet/nodeconfig/controller.go

```diff
@@ -0,0 +1,270 @@
+		return cc.defaultConfig
+	} // Assert: we will not use the init or default configurations, unless we roll back to lkg; curUID is a real UID
+
+	// check whether the current config is marked bad
+	if bad, entry := cc.isBadConfig(curUID); bad {
+		infof("current config %q was marked bad for reason %q at time %q", curUID, entry.reason, entry.time)
+		return cc.lkgRollback(entry.reason, apiv1.ConditionFalse)
+	}
+
+	// TODO(mtaufen): consider re-verifying integrity and re-attempting download when a load/verify/parse/validate
+	// error happens outside trial period, we already made it past the trial so it's probably filesystem corruption
+	// or something else scary
+
+	// load the current config
+	toVerify, err := cc.loadCheckpoint(curSymlink)
+	if err != nil {
+		// TODO(mtaufen): rollback and mark bad for now, but this could reasonably be handled by re-attempting a download,
+		// it probably indicates some sort of corruption
+		return cc.badRollback(curUID, fmt.Sprintf("failed to load current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// verify the integrity of the configuration we just loaded
+	toParse, err := toVerify.verify()
+	if err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to verify current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// parse the configuration we just loaded into a KubeletConfiguration
+	cur, err := toParse.parse()
+	if err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to parse current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// validate current config
+	if err := validateConfig(cur); err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to validate current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// check for crash loops if we're still in the trial period
+	if cc.curInTrial(cur.ConfigTrialDuration.Duration) {
+		if cc.curCrashLooping(*cur.CrashLoopThreshold) {
+			return cc.badRollback(curUID, fmt.Sprintf("current failed trial period due to crash loop (UID %q)", curUID), "")
+		}
+	} else if !cc.curIsLkg() {
+		// when the trial period is over, the current config becomes the last-known-good
+		cc.setCurAsLkg()
+	}
+
+	// update the status to note that we will use the current config
+	cc.setConfigOK(fmt.Sprintf("using current (UID: %q)", curUID), "passed all checks", apiv1.ConditionTrue)
+
+	return cur
+}
+
+// initSync launches the sync loop that watches the API server for new configuration
+func (cc *NodeConfigController) initSync() {
+	if cc.client != nil {
+		infof("starting sync loop")
+		go func() {
+			fieldselector := fmt.Sprintf("metadata.name=%s", cc.nodename)
+
+			// Add some randomness to resync period, which can help avoid controllers falling into lock-step
+			minResyncPeriod := 30 * time.Second
+			factor := rand.Float64() + 1
+			resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+
+			_, ctl := cache.NewInformer(
```

Comment:
Does that get filtered in the client or on the API server? If it happens in the client... it won't scale to have every Kubelet get pinged about updates to every node in the cluster.